src/variety/bosnianroad.js

```diff
@@ -0,0 +1,319 @@
+				return false;
+			}
+			if (this.qcmp === 1) {
+				return true;
+			}
+			if (!this.puzzle.execConfig("autocmp")) {
+				return false;
+			}
+			return this.checkComplete();
+		},
+		
+		checkComplete: function() {
+			if (!this.isValidNum()) {
+				return true;
+			}
+
+			var clist = this.getClist();
+
+			var undecided = clist.some(function(cc) {
+				return cc.qans === 0 && cc.qsub === 0;
+			});
+			if (undecided) {
+				return false;
+			}
+
+			var cnt = clist.filter(function(cc) {
+				return cc.isShade();
+			}).length;
+			return cnt === this.qnum;
+		}
+	},
+	
+	Border: {
+		isBorder: function() {
+			return (this.sidecell[0].qnum === -1) !== (this.sidecell[1].qnum === -1);
+		}
+	},
+
+	Board: {
+		hasborder: 1
+	},
+
+	AreaShadeGraph: {
+		enabled: true,
+		coloring: true
+	},
+
+	Graphic: {
+		autocmp: "number",
+		qcmpcolor: "rgb(144,144,144)",
+
+		setRange: function(x1, y1, x2, y2) {
+			var puzzle = this.puzzle,
+				bd = puzzle.board;
+			if (puzzle.execConfig("autocmp")) {
+				x1 = bd.minbx - 2;
+				y1 = bd.minby - 2;
+				x2 = bd.maxbx + 2;
+				y2 = bd.maxby + 2;
+			}
+			this.common.setRange.call(this, x1, y1, x2, y2);
+		},
+		
+		cluebgcolor: "rgb(224,224,224)",
+		hideHatena: true,
+		irowakeblk: true,
+
+		getBGCellColor: function(cell) {
+			if (cell.qnum !== -1) {
+				return this.cluebgcolor;
+			}
+			return this.getBGCellColor_error1(cell);
+		},
+		
+		getQuesNumberColor: function(cell) {
+			var qnum_color = this.getQuesNumberColor_qnum(cell);
+			if ((cell.error || cell.qinfo) === 1) {
+				return qnum_color;
+			}
+			if (cell.isCmp()) {
+				return this.qcmpcolor;
+			}
+			return qnum_color;
+		},
+
+		paint: function() {
+			this.drawBGCells();
+			this.drawDotCells();
+
+			this.drawShadedCells();
+			this.drawGrid();
+			this.drawGridGaps();
+			this.drawBorders();
+
+			this.drawQuesNumbers();
+			
+			this.drawPekes();
+
+			this.drawChassis();
+
+			this.drawTarget();
+		},
+
+		drawGridGaps: function() {
+			var g = this.vinc("gridgaps", "crispEdges", true);
+			var i,
+				bd = this.board,
+				bw = this.bw,
+				bh = this.bh;
+			g.lineWidth = 1;
+			for (i = 0; i < bd.border.length; i++) {
+				var border = bd.border[i];
+				g.vid = "bdgap_" + i;
+				if (border.sidecell[0].qnum !== -1 && border.sidecell[1].qnum !== -1) {
+					var bx1 = border.sidecross[0].bx,
+						by1 = border.sidecross[0].by,
+						bx2 = border.sidecross[1].bx,
+						by2 = border.sidecross[1].by;
+					g.strokeStyle = this.cluebgcolor;
+					g.strokeLine(bx1 * bw, by1 * bh, bx2 * bw, by2 * bh);
+				} else {
+					g.vhide();
+				}
+			}
+		}
+	},
+
+	Encode: {
+		decodePzpr: function(type) {
+			this.decodeNumber10();
+		},
+		encodePzpr: function(type) {
+			this.encodeNumber10();
+		}
+	},
+
+	FileIO: {
+		decodeData: function() {
+			this.decodeCellQnum();
+			this.decodeCellAns();
+		},
+		encodeData: function() {
+			this.encodeCellQnum();
+			this.encodeCellAns();
+		}
+	},
+
+	AnsCheck: {
```

Comment:
Could you add a 2x2 shaded check like in Snake? While it's technically redundant, in practice it's nicer to have a different error message stating that the loop goes back on itself.

Reply:
It's somewhat underspecified but, as best as I can tell, unlike in Snake, a shaded 2x2 with nothing else in the grid is technically a valid Bosnian Road solution. It's admitted as such under the current implementation and I find it easier to argue for than against it. Perhaps I can make the 2x2 check only trigger if the connected component has more than 4 shaded cells?

Reply:
> Perhaps I can make the 2x2 check only trigger if the connected component has more than 4 shaded cells?

That's a good idea.